### PR TITLE
Add TransactionNameProvider to allow the name definition from Unknown transactions on ASP.NET Core

### PR DESCRIPTION
**Pull request by lucas-zimerman**
- The idea is to allow the user to set the transaction name on cases where the SDK isn't able to define the current Transaction name.

- I'm also altering the GCP Integration since it already uses the integration, but without relying on the HttpContext to set the name.

Solves #1373.

### CODE REVIEW

File: src/Sentry.AspNetCore/SentryAspNetCoreOptions.cs [hunk 1]
```diff
@@ -43,6 +43,12 @@ public class SentryAspNetCoreOptions : SentryLoggingOptions
     /// </summary>
     public TimeSpan FlushTimeout { get; set; } = TimeSpan.FromSeconds(2);
 
+    /// <summary>
+    /// RouteName sets the transaction name that was created by the ASPNetCore integration
+    /// when the integration isn't able to define the transaction name..
+    /// </summary>
+    public ISentryRouteName? RouteName { get; set; }
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/SentryAspNetCoreOptions.cs:50]: Something that better reflects what this options does:

```suggestion
    public ITransactionNameProvider? TransactionNameProvider { get; set; }
```

File: src/Sentry.AspNetCore/SentryAspNetCoreOptions.cs [hunk 2]
```diff
@@ -43,6 +43,12 @@ public class SentryAspNetCoreOptions : SentryLoggingOptions
     /// </summary>
     public TimeSpan FlushTimeout { get; set; } = TimeSpan.FromSeconds(2);
 
+    /// <summary>
+    /// RouteName sets the transaction name that was created by the ASPNetCore integration
+    /// when the integration isn't able to define the transaction name..
+    /// </summary>
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/SentryAspNetCoreOptions.cs:49]:
```suggestion
    /// <summary>
    /// The strategy to define the name of a transaction based on the HttpContext
    /// </summary>
    /// <remarks >
    /// The SDK can name transactions automatically when using MVC or Endpoint Routing. In other cases, like when serving static files, it fallback to Unknown Route. This hook allows custom code to define a transaction name given a HttpContext.
    /// </remarks>
```

File: src/Sentry.AspNetCore/SentryTracingMiddleware.cs [hunk 3]
```diff
@@ -110,6 +110,11 @@ public async Task InvokeAsync(HttpContext context)
             var transaction = TryStartTransaction(context);
             var initialName = transaction?.Name;
 
+            if (_options.RouteName is { } route)
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/SentryTracingMiddleware.cs:113]: this needs to happen earlier

**Review by SimonCropp — APPROVED** [review]
LGTM. i pushed some very small cosmetic changes

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 ## Features
 
+- Add ITransactionNameProvider to allow the name definition from Unknown transactions on ASP.NET Core ([#1421](https://github.com/getsentry/sentry-dotnet/pull/1421))
```
Comment (bruno-garcia) [inline CHANGELOG.md:7]: I guess this needs to be reworded now

File: src/Sentry.AspNetCore/ITransactionNameProvider.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+using Microsoft.AspNetCore.Http;
```
Comment (lucas-zimerman) [inline src/Sentry.AspNetCore/ITransactionNameProvider.cs:1]: Filename needs to be renamed.

**Comment by SimonCropp** [comment]
do we need a corresponding docs PR?

**Comment by lucas-zimerman** [comment]
> do we need a corresponding docs PR?

We deffo need some docs but I believe we could work on it once we have a released version with this PR bundled.

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**